lib/ffi-icu/locale.rb

```diff
@@ -96,6 +101,16 @@ def display_name(locale = nil)
       end
     end
 
+    def display_name_with_context(locale, contexts = [])
+      contexts = DISPLAY_CONTEXT.select { |context| contexts.include?(context) }.values
+      pointer = FFI::MemoryPointer.new(:int, contexts.length).write_array_of_int(contexts)
+      open_for_context = ICU::Lib.check_error { |status| ICU::Lib.uldn_openForContext(@id, pointer, contexts.length, status) }
```

Comment:
You get a [LocaleDisplayNames](https://unicode-org.github.io/icu-docs/apidoc/dev/icu4c/uldnames_8h.html#a090a6edf8fb5f22aec739692187a0014).

```suggestion
      locale_display_names = ICU::Lib.check_error { |status| ICU::Lib.uldn_openForContext(@id, pointer, contexts.length, status) }
```

Since you are creating an instance, you also have to close it. Otherwise, you will leak memory. You can implement `uldn_close`.

You can implement wrap `locale_display_names` with additional functions with a function as well so you can use ensure block.

```ruby
def with_uldp(locale_id, contexts) # `uldp` is probably a really bad name
  pointer = FFI::MemoryPointer.new(:int, contexts.length).write_array_of_int(contexts)
  locale_display_names = ICU::Lib.check_error { |status| ICU::Lib.uldn_openForContext(locale_id, pointer, contexts.length, status) }
  yield locale_display_names
ensure
  ICU::Lib.uldn_close(locale_display_names) if locale_display_names
end

def display_name_with_context(locale, contexts = [])
  contexts = DISPLAY_CONTEXT.select { |context| contexts.include?(context) }.values
  with_uldp(@id, contexts) do |locale_display_names|
    ICU::Lib::Util.read_uchar_buffer(256) do |buffer, status|
      ICU::Lib.uldn_localeDisplayName(locale_display_names, locale, buffer, buffer.size, status)
    end
  end
end
```